File: 08-Recursion-Backtracking/03-Permutations/cpp/permutations.cpp
/*
 * Problem: Permutations
 * Source:  LeetCode 46
 * Approach: Swap-based backtracking.
 * Time: O(n! * n), Space: O(n)
 * Tags: backtracking, permutations
 */
#include <bits/stdc++.h>
using namespace std;

void permute(vector<int>& nums, int idx, vector<vector<int>>& result) {
    if (idx == nums.size()) { result.push_back(nums); return; }
    for (int i = idx; i < nums.size(); i++) {
        swap(nums[idx], nums[i]);
        permute(nums, idx + 1, result);
        swap(nums[idx], nums[i]);
    }
}

int main() {
    vector<int> nums = {1, 2, 3};
    vector<vector<int>> result;
    permute(nums, 0, result);
    for (auto& p : result) { for (int x : p) cout << x << " "; cout << endl; }
    return 0;
}

File: 08-Recursion-Backtracking/03-Permutations/java/Permutations.java
import java.util.*;
public class Permutations {
    public static List<List<Integer>> permute(int[] nums) {
        List<List<Integer>> result = new ArrayList<>();
        backtrack(nums, 0, result);
        return result;
    }
    static void backtrack(int[] nums, int idx, List<List<Integer>> result) {
        if (idx == nums.length) {
            List<Integer> perm = new ArrayList<>();
            for (int x : nums) perm.add(x);
            result.add(perm); return;
        }
        for (int i = idx; i < nums.length; i++) {
            int t = nums[idx]; nums[idx] = nums[i]; nums[i] = t;
            backtrack(nums, idx + 1, result);
            t = nums[idx]; nums[idx] = nums[i]; nums[i] = t;
        }
    }
    public static void main(String[] args) { System.out.println(permute(new int[]{1, 2, 3})); }
}
